compiler-core/src/erlang/tests/type_params.rs

```diff
@@ -10,3 +10,36 @@ fn named_args_count_once() {
         "
     );
 }
+
+#[test]
+fn nested_named_args_count_once() {
+    assert_erl!(
+        "
+        pub type Nested(a, b) {
+            VariantX(field_a: a, field_b: b)
+            VariantY(field_a: a, field_b: b)
+        }
+
+        pub type Parent(a, b) {
+            ParentX(Nested(a,b))
+            ParentY(Nested(a,b))
+        }
+
+        pub fn foo1() -> Parent(a, a) {
+            todo
+        }
+
+        pub fn foo2() -> Parent(a, b) {
+            todo
+        }
+
+        pub fn foo3(val_a: a) -> Parent(a, a) {
+            todo
+        }
+
+        pub fn foo4(val_a: a, val_b: b) -> Parent(a, b) {
+            todo
+        }
+        "
+    );
+}
```

Comment:
This test does many things. Could you break it out into one test per example please? Thanks

I can't see any nested types that use the same type variables at different levels- I think that's where we may have the issue.

Reply:
> This test does many things. Could you break it out into one test per example please? Thanks

Sure!


> I can't see any nested types that use the same type variables at different levels- I think that's where we may have the issue.

Can you give an example of such a type? Then I can see what is generated now vs before this PR.

Reply:
Maybe something like `Wibble(a, Wibble(a, b))`?

Reply:
Sure!

For types that have two or more variants, I've simplified the test cases to:
- `Result(a, a)` generates `{ok, any()} | {error, any()}`
- `Result(a, Result(a, b))` generates `{ok, any()} | {error, {ok, any()} | {error, any()}}`
- `#(a, Result(a, b))` generates `{I, {ok, I} | {error, any()}}`

I think this is correct?

However, for the case with a single variant, the current patch generates sub-optimal code:
- `Wibble(a, a)` generates `wibble(any(), any())`, while it technically could be `wibble(A, A)`
- `Wibble(a, Wibble(a, b))` generates `wibble(any(), wibble(any(), any()))`, while it could be `wibble(A, wibble(A, any()))`
- `#(a, Wibble(a, b))` generates `{A, wibble(A, any())}`

On the other hand, we've seen from the discussion on https://github.com/erlang/otp/issues/8533 that there's not really a way to express that the two a type variables are the same type, so not sure if we want to generate types like that?

In that case, this PR should take a different solution I think and take into account the `constructors` while counting the parameter occurrences ([here](https://github.com/gleam-lang/gleam/blob/main/compiler-core/src/erlang.rs#L265)).

Reply:
Now that I think about it, shouldn’t uniquely counting across constructors be the only thing we need? (for custom types)